### PR TITLE
Feature: Unique Pinned Posts

### PR DESCRIPTION
**Pull request by efuller**
Closes ticket #48 

## Description
Provides functionality that will allow for unique, pinned posts on a page. This means that if a post is pinned, it will not display anywhere else on the page.

## Notes
- Uniqued pinned posts relies on deduplication being enabled (the editor toggle should reflect this)
- Will bump version after CR

## Additional Updates
- GH action workflows have been updated to allow for tests to be run properly
- Mantle Testkit has been updated

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -54,7 +58,7 @@
   "prefer-stable": true,
   "scripts": {
     "phpcbf": "phpcbf .",
-    "phpcs": "phpcs .",
+    "phpcs": "phpcs . -sn",
```
Comment (kevinfodness) [inline composer.json]: We shouldn't define these flags here. The `s` flag is already defined in `.phpcs.xml` and we don't want to suppress warnings.

File: entries/slotfills/deduplication/deduplication.tsx [hunk 2]
```diff
@@ -2,23 +2,33 @@ import { ToggleControl } from '@wordpress/components';
...
 import { usePostMetaValue } from '@alleyinteractive/block-editor-tools';
-
-/**
- * The following code is a temporary fix.
- *
- * Once the issue linked below is resolved, this code can be removed.
- * @link https://github.com/alleyinteractive/alley-scripts/issues/473
- */
 import { useSelect } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
```
Comment (kevinfodness) [inline entries/slotfills/deduplication/deduplication.tsx]: We should import from React instead. The WP project has transitioned away from using `@wordpress/element` to using React directly.

File: services/deduplicate/index.ts [hunk 3]
```diff
@@ -96,26 +96,28 @@ export function mainDedupe() {
...
-  // });
+
+  /**
+   * This block of code is responsible for enforcing the unique pinned posts setting in the editor.
+   */
+  if (wpCurateUniquePinnedPosts) {
+    queryBlocks.forEach((queryBlock) => {
+      const { attributes } = queryBlock;
+      const { posts } = attributes;
+      posts?.forEach((post) => {
```
Comment (kevinfodness) [inline services/deduplicate/index.ts]: We can collapse this quite a bit:
```suggestion
    queryBlocks.forEach((queryBlock) => {
      queryBlock?.attributes?.posts?.forEach((post) => {
```

File: src/class-supported-post-types.php [hunk 4]
```diff
@@ -92,6 +92,14 @@ public function register_post_meta(): void {
 				'type' => 'boolean',
 			]
 		);
+		register_meta_helper(
+			'post',
+			$this->get_supported_post_types(),
+			'wp_curate_unique_pinned_posts',
+			[
+				'type' => 'boolean',
```
Comment (kevinfodness) [inline src/class-supported-post-types.php]: We should set a default of `false`

File: src/post-ids/class-pinned-in-post-content.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		if ( $main_query->is_singular() ) {
+			$post = $main_query->get_queried_object();
+
+			if ( $post instanceof \WP_Post ) {
+				// Unique pinned posts relies on deduplication being enabled.
+				$post_level_unique         = get_post_meta( $post->ID, 'wp_curate_unique_pinned_posts', true );
+				$post_level_deduplication  = get_post_meta( $post->ID, 'wp_curate_deduplication', true );
+				$post_level_unique_enabled = $post_level_unique && $post_level_deduplication;
+
+				if ( true === (bool) $post_level_unique_enabled ) {
```
Comment (kevinfodness) [inline src/post-ids/class-pinned-in-post-content.php]: No need to do several casts to boolean here, the `&&` will do it:
```suggestion
				if ( $post_level_unique && $post_level_deduplication ) {
```

File: tests/feature/UniquePinnedPostTest.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	 * Run before each test.
+	 */
+	protected function setUp(): void {
+		parent::setUp();
+
+		$this->set_permalink_structure( '/%postname%/' );
+
+		static::factory()->post->create(
+			[
+				'post_title' => 'TheWrapTestPost 1',
```
Comment (kevinfodness) [inline tests/feature/UniquePinnedPostTest.php]: We shouldn't have client names in tests in open source repos. Let's just make this "Test Post 1" etc for others in this file.

**Review by kevinfodness — CHANGES_REQUESTED** [review]

**Comment by efuller** [comment]
@kevinfodness I've adjusted code per your review and requested another review. Could you have a look when you have a moment?

**Review by kevinfodness — APPROVED** [review]
🍣

**Pull request merged**